### PR TITLE
Switch to focus mode if read-only list items and their list parents are focusable and gain focus

### LINKED ISSUE #13221: NVDA doesn't toggle browse mode off when focus is moved to an interactive list
I believe I have discovered a regression in NVDA. @jcsteh can correct me if I am misremembering, but I recall NVDA _used_ to toggle browse mode off if focus was moved to an HTML list made interactive with `tabindex` and keyboard shortcuts implemented in JavaScript. Reason I remember this: while I was working at Twitter I made the timeline an interactive list so that customers could use keyboard shortcuts provided by Twitter to navigate their home timeline. And I recall working with @jcsteh on ensuring this worked in NVDA. Interestingly, the same code example I provided in this bug report works fine in JAWS in that it toggles browse mode off when focus lands on an ARIA `listitem` instance.

### Steps to reproduce:

1. Start NVDA
2. Open https://codepen.io/toddlr/pen/mdBjoJV
3. Use tab-key navigation to move focus to the first ARIA `listitem` in the ARIA `list`

### Actual behavior:
NVDA doesn't toggle browse mode off

### Expected behavior:
NVDA should toggle browse mode off so the up/down arrow keys can be used to navigate (move focus) between listitems in the list

#### Does the issue still occur after restarting your computer?
Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
No

#### If NVDA add-ons are disabled, is your problem still occurring?
I'm not using any add-ons

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes

### PR DESCRIPTION
**Pull request by MarcoZehe**
### Link to issue number:

Fixes #13221

### Summary of the issue:

In the Microsoft Edge Downloads manager (CTRL+J), as well as in the channel and DM outputs of the Slack app, HTML lists and list items are used to display the items. These are made focusable via either aria-activedescendant, or the rowing tabIndex pattern. The reason is that for compatibility, these can be used to host rich content, such as whole sub documents or controls. Until now, NVDA would not automatically switch to focus mode when focus lands on such a focusable list item.

### Description of how this pull request fixes the issue:

This PR adds a rule for enabling passthrough mode if a list item that is read-only, has state focused, and at the same time its paarent is of role list and is focusable.

### Testing strategy:

1. Run NVDA.
2. Go into Microsoft Edge.
3. Press Control+J to invoke Downloads Manager.
4. Focus should be on the first downloaded item, and passthrough mode should automatically be enabled.

### Known issues with pull request:

None known.

### Change log entries:
New features
Changes
Bug fixes

* In the Microsoft Edge downloads manager, NVDA will now automatically switch to focus mode once the list item with the most recent download gains focus. #13221 

For Developers

### Code Review Checklist:

- [X] Pull Request description:
  - description is up to date
  - change log entries
- [X] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [X] API is compatible with existing add-ons.
- [ ] Documentation: Not needed since expected behavior is introduced where user would previously have to manually switch pass-through mode on.
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [X] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

### CODE REVIEW

**Comment by TheQuinbox** [comment]
Nice, looks good to me. In NVDA changelog entries though, it's typical to link to the issue that it solves (if applicable).

**Comment by MarcoZehe** [comment]
Thanks @TheQuinbox, I just updated the initial description with that, as well as the checkbox for the testing part, since the manual check checks out nicely.

**Comment by CyrilleB79** [comment]
Typo to fix in the change log:
#13321 -> #13221

**Comment by lukaszgo1** [comment]
Have you considered adding Chrome system test demonstrating the behavior based on the code snippet in #13221?

**Comment by MarcoZehe** [comment]
Thanks @CyrilleB79, corrected the typo.

@lukaszgo1 I'll try that system test thing, but this is all still very confusing, so may take a while. And it's made difficult by the fact that I can't currently even run local builds, and that includes tests. See my message in the nvda-devel mailing list. But I'll try nevertheless.

**Comment by leonardder** [comment]
Thanks Marco, this is a nice improvement! One question: Isn't it just enough that a list item has focus? Why check the focusable state on the parent list as well??

**Comment by MarcoZehe** [comment]
@leonardder I was going with @michaelDCurran's statement from https://github.com/nvaccess/nvda/issues/13221#issuecomment-1039904489:

> We will switch to focus mode if a listItem gets focus, but only if the list itself is also focusable. I.e. the list has at least tabindex of -1.

This is how MS Edge's Download Manager list works, and it means that @kloots will have to make a slight adjustment in Slack to put tabindex="-1" on the actual list, which Slack currently doesn't do yet.

**Comment by michaelDCurran** [comment]
For reference, the reasoning behind requiring focusable on the list is to limit where this code will apply.
1. Making ul (or role="list) and (li or role="listitem") interactive like this does not quite fit standards as really these are presentational read-only lists. Interactive lists should be role="listbox" and role="option".
2. There may already exist some web pages that deliberately set focus to a presentational list item when the page is scrolled to an arbitrary location. In this case switching to focus mode would be annoying and inappropriate.

But, it has been made clear that the standard does not really fit some particular use cases, and therefore this change is needed. Therefore it is being added, but with the limitation that the list must also be focusable. This idea was proposed by Slack themselves.

File: tests/system/robot/chromeTests.robot [hunk 1]
```diff
@@ -120,3 +120,6 @@ Quick Nav reports target first
 Focus reports target first
 	[Documentation]	Focus target should always be reported before ancestors. Ancestors should be reported from inner to outer.
 	test_focusTargetReporting
+focus mode is turned on on focused read-only list item
+	[Documentation]Focused list items with a focusable list container should cause focus mode to be turned on automatically.
```
Comment (michaelDCurran) [inline tests/system/robot/chromeTests.robot:124]: Missing tab character after `[Documentation]`
```suggestion
	[Documentation]	Focused list items with a focusable list container should cause focus mode to be turned on automatically.
```

**Review by michaelDCurran — CHANGES_REQUESTED** [review]

**Comment by michaelDCurran** [comment]
I think the title of this pr should refer to Focus mode, not Browse mode.

**Comment by MarcoZehe** [comment]
> I think the title of this pr should refer to Focus mode, not Browse mode.

Done.

**Comment by MarcoZehe** [comment]
Sorry this is going so slowly. I cannot run the tests on my local machine because it is not an English language Windows, and therefore all NVDA tests fall on their face. So I have to peel this like an onion.

**Comment by MarcoZehe** [comment]
@michaelDCurran I don't understand this error message. The previous test failures were clear, but this one confuses me. It says:

> Error message:
focus mode - focus link in document and turn off focus mode
speech Actual != Expected: message
Todd Kloots  link !=   message
Todd Kloots  link
Browse mode

Any help is appreciated.

**Comment by MarcoZehe** [comment]
@michaelDCurran I went with your suggestion of simplifying the test case after all, and it now passes. I can't clear the "changes requested" flag from the issue even though I applied your suggested code change, and also updated the PR title. So, tossing this back for review in hopes that things look good so far. I'd also be happy to help out with #12896 if that's desired.

**Comment by michaelDCurran** [comment]
Thanks for your work on this @MarcoZehe.
I shall review this  sometime in the next couple of days.

**Comment by brunoprietog** [comment]
This is great! Can you think of a way to fix #12266? It's a similar list focus problem. Thanks!

**Comment by MarcoZehe** [comment]
I just commented on the issue: It is a different bug, which will not be fixed by this pull request. The bug is in the Chromium engine, not NVDA in that case. The markup is also radically different from what is being used here.

**Review by michaelDCurran — APPROVED** [review]

**Comment by kloots** [comment]
@michaelDCurran I see this has been added to the 2022.1 release milestone. Does the team have a release date spec'd for 2022.1?

**Comment by michaelDCurran** [comment]
@kloots  perhaps end of March / early April.

**Comment by kloots** [comment]
@michaelDCurran curious about the release date for this fix: any update as to when this will go out to production?

**Comment by michaelDCurran** [comment]
It is currently in NVDA 2022.1beta3. There are plans for a beta4 in the 
coming week or so, and then the release will most likely be 3 or so 
weeks after that. So rough guess is mid May.

**Comment by kloots** [comment]
Great. Thanks @michaelDCurran. Dumb question: how do I get the beta?

**Comment by michaelDCurran** [comment]
Link to beta3 post: https://www.nvaccess.org/post/nvda-2022-1beta3/

**Comment by kloots** [comment]
Thanks @michaelDCurran. One last question, do you have a sense for how quickly NVDA users upgrade?

**Comment by michaelDCurran** [comment]
Users upgrade pretty slowly I'm sorry to say.
NVDA 2021.3.5 has been out for about 3 weeks now, and only 30% of users 
have upgraded. The rest are all on random older versions. The next most 
popular version is 2021.3.1 at just under 10%.

We alert users of updates within NVDA when ever they are available.

**Pull request merged**